From 58da34ca01573f47a4beb806e10cd18898f09a46 Mon Sep 17 00:00:00 2001
From: Jordan Lazare <Jordan.Lazare@amd.com>
Date: Mon, 25 Apr 2016 13:54:07 -0400
Subject: [PATCH 0453/1722] drm/amd/dal: Only create up to the max number of
 targets for asic

This change limits the number of crtc's reported to drm to the
maximum number of displays that can be driven.

This fixes pageflip/validation issues when trying to drive
more displays than capable in the hardware. Usermode continously
failed validation since it keeps trying to lower the timing to fit
within the max bandwidth, but validation will always fail in this case
since it fails due to not having enough controllers, not due to
bandwidth restraints

Signed-off-by: Jordan Lazare <Jordan.Lazare@amd.com>
Acked-by: Harry Wentland <harry.wentland@amd.com>
Signed-off-by: Alex Deucher <alexander.deucher@amd.com>
Signed-off-by: Kalyan Alle <kalyan.alle@amd.com>
---
 drivers/gpu/drm/amd/dal/dc/core/dc.c                |  4 +++-
 drivers/gpu/drm/amd/dal/dc/dce110/dce110_resource.c | 13 +++++++++++--
 drivers/gpu/drm/amd/dal/dc/dce112/dce112_resource.c | 14 +++++++++++---
 drivers/gpu/drm/amd/dal/dc/dce80/dce80_resource.c   | 13 +++++++++++--
 drivers/gpu/drm/amd/dal/dc/inc/core_types.h         |  8 ++++----
 5 files changed, 40 insertions(+), 12 deletions(-)

diff --git a/drivers/gpu/drm/amd/dal/dc/core/dc.c b/drivers/gpu/drm/amd/dal/dc/core/dc.c
index 662b33d..1fada66 100644
--- a/drivers/gpu/drm/amd/dal/dc/core/dc.c
+++ b/drivers/gpu/drm/amd/dal/dc/core/dc.c
@@ -416,7 +416,9 @@ struct dc *dc_create(const struct dc_init_data *init_params)
 	/*TODO: separate HW and SW initialization*/
 	core_dc->hwss.init_hw(core_dc);
 
-	core_dc->public.caps.max_targets = core_dc->res_pool.pipe_count;
+	core_dc->public.caps.max_targets = dm_min(
+			core_dc->res_pool.pipe_count,
+			core_dc->res_pool.stream_enc_count);
 	core_dc->public.caps.max_links = core_dc->link_count;
 	core_dc->public.caps.max_audios = core_dc->res_pool.audio_count;
 
diff --git a/drivers/gpu/drm/amd/dal/dc/dce110/dce110_resource.c b/drivers/gpu/drm/amd/dal/dc/dce110/dce110_resource.c
index b7d485b..a06002b 100644
--- a/drivers/gpu/drm/amd/dal/dc/dce110/dce110_resource.c
+++ b/drivers/gpu/drm/amd/dal/dc/dce110/dce110_resource.c
@@ -925,6 +925,17 @@ bool dce110_construct_resource_pool(
 	pool->adapter_srv = as;
 	pool->funcs = &dce110_res_pool_funcs;
 
+	/*************************************************
+	 *  Resource + asic cap harcoding                *
+	 *************************************************/
+	pool->pipe_count = 4;
+	pool->stream_enc_count = 3;
+
+	/*************************************************
+	 *  Create resources                             *
+	 *************************************************/
+
+
 	pool->stream_engines.engine.ENGINE_ID_DIGA = 1;
 	pool->stream_engines.engine.ENGINE_ID_DIGB = 1;
 	pool->stream_engines.engine.ENGINE_ID_DIGC = 1;
@@ -982,8 +993,6 @@ bool dce110_construct_resource_pool(
 
 	}
 
-	pool->pipe_count = dal_adapter_service_get_func_controllers_num(as);
-	pool->stream_enc_count = dal_adapter_service_get_stream_engines_num(as);
 	pool->scaler_filter = dal_scaler_filter_create(ctx);
 	if (pool->scaler_filter == NULL) {
 		BREAK_TO_DEBUGGER();
diff --git a/drivers/gpu/drm/amd/dal/dc/dce112/dce112_resource.c b/drivers/gpu/drm/amd/dal/dc/dce112/dce112_resource.c
index f9441c1..3ee692c 100644
--- a/drivers/gpu/drm/amd/dal/dc/dce112/dce112_resource.c
+++ b/drivers/gpu/drm/amd/dal/dc/dce112/dce112_resource.c
@@ -1027,6 +1027,17 @@ bool dce112_construct_resource_pool(
 	pool->adapter_srv = adapter_serv;
 	pool->funcs = &dce112_res_pool_funcs;
 
+	/*************************************************
+	 *  Resource + asic cap harcoding                *
+	 *************************************************/
+	pool->pipe_count =
+		dal_adapter_service_get_func_controllers_num(adapter_serv);
+	pool->stream_enc_count = dal_adapter_service_get_stream_engines_num(adapter_serv);
+
+	/*************************************************
+	 *  Create resources                             *
+	 *************************************************/
+
 	pool->stream_engines.engine.ENGINE_ID_DIGA = 1;
 	pool->stream_engines.engine.ENGINE_ID_DIGB = 1;
 	pool->stream_engines.engine.ENGINE_ID_DIGC = 1;
@@ -1085,9 +1096,6 @@ bool dce112_construct_resource_pool(
 
 	}
 
-	pool->pipe_count =
-		dal_adapter_service_get_func_controllers_num(adapter_serv);
-	pool->stream_enc_count = dal_adapter_service_get_stream_engines_num(adapter_serv);
 	pool->scaler_filter = dal_scaler_filter_create(ctx);
 	if (pool->scaler_filter == NULL) {
 		BREAK_TO_DEBUGGER();
diff --git a/drivers/gpu/drm/amd/dal/dc/dce80/dce80_resource.c b/drivers/gpu/drm/amd/dal/dc/dce80/dce80_resource.c
index 6a56f65..2de7aa0 100644
--- a/drivers/gpu/drm/amd/dal/dc/dce80/dce80_resource.c
+++ b/drivers/gpu/drm/amd/dal/dc/dce80/dce80_resource.c
@@ -688,6 +688,17 @@ bool dce80_construct_resource_pool(
 	pool->adapter_srv = as;
 	pool->funcs = &dce80_res_pool_funcs;
 
+
+	/*************************************************
+	 *  Resource + asic cap harcoding                *
+	 *************************************************/
+	pool->pipe_count = dal_adapter_service_get_func_controllers_num(as);
+	pool->stream_enc_count = dal_adapter_service_get_stream_engines_num(as);
+
+	/*************************************************
+	 *  Create resources                             *
+	 *************************************************/
+
 	pool->stream_engines.engine.ENGINE_ID_DIGA = 1;
 	pool->stream_engines.engine.ENGINE_ID_DIGB = 1;
 	pool->stream_engines.engine.ENGINE_ID_DIGC = 1;
@@ -753,8 +764,6 @@ bool dce80_construct_resource_pool(
 
 	}
 
-	pool->pipe_count = dal_adapter_service_get_func_controllers_num(as);
-	pool->stream_enc_count = dal_adapter_service_get_stream_engines_num(as);
 	pool->scaler_filter = dal_scaler_filter_create(ctx);
 	if (pool->scaler_filter == NULL) {
 		BREAK_TO_DEBUGGER();
diff --git a/drivers/gpu/drm/amd/dal/dc/inc/core_types.h b/drivers/gpu/drm/amd/dal/dc/inc/core_types.h
index c13c224..86edf4d 100644
--- a/drivers/gpu/drm/amd/dal/dc/inc/core_types.h
+++ b/drivers/gpu/drm/amd/dal/dc/inc/core_types.h
@@ -239,8 +239,8 @@ struct resource_pool {
 	struct timing_generator *timing_generators[MAX_PIPES];
 	struct stream_encoder *stream_enc[MAX_PIPES * 2];
 
-	uint8_t pipe_count;
-	uint8_t stream_enc_count;
+	unsigned int pipe_count;
+	unsigned int stream_enc_count;
 
 	union supported_stream_engines stream_engines;
 
@@ -250,10 +250,10 @@ struct resource_pool {
 	struct clock_source *dp_clock_source;
 
 	struct clock_source *clock_sources[MAX_CLOCK_SOURCES];
-	uint8_t clk_src_count;
+	unsigned int clk_src_count;
 
 	struct audio *audios[MAX_PIPES];
-	uint8_t audio_count;
+	unsigned int audio_count;
 
 	struct display_clock *display_clock;
 	struct adapter_service *adapter_srv;
-- 
2.7.4

